internal/kms/azure_vault.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	err = setConfigString(&kms.clientCertificate, secrets, azureClientCertificate)
+	if err != nil {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+func (kms *azureKeyVaultKMS) Destroy() {
+	// Nothing to do.
+}
+
+func (kms *azureKeyVaultKMS) getService() (*azsecrets.Client, error) {
+	certs, key, err := azidentity.ParseCertificates([]byte(kms.clientCertificate), []byte{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Azure client certificate: %w", err)
+	}
+	creds, err := azidentity.NewClientCertificateCredential(kms.tenantID, kms.clientID, certs, key, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Azure credentials: %w", err)
+	}
+
+	azClient, err := azsecrets.NewClient(kms.vaultURL, creds, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Azure client: %w", err)
+	}
+
+	return azClient, nil
+}
+
+func (kms *azureKeyVaultKMS) getSecrets() (map[string]interface{}, error) {
+	c, err := k8s.NewK8sClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to kubernetes to "+
+			"get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	secret, err := c.CoreV1().Secrets(kms.namespace).Get(context.TODO(),
+		kms.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
+	for k, v := range secret.Data {
+		switch k {
+		case azureClientCertificate:
+			config[k] = string(v)
+		default:
+			return nil, fmt.Errorf("unsupported option for KMS provider %q: %s", kmsTypeAzure, k)
+		}
+	}
+
+	return config, nil
+}
+
+// FetchDEK returns passphrase from Azure key vault.
+func (kms *azureKeyVaultKMS) FetchDEK(key string) (string, error) {
+	svc, err := kms.getService()
+	if err != nil {
+		return "", fmt.Errorf("failed to get KMS service: %w", err)
+	}
+
+	getSecretResponse, err := svc.GetSecret(context.TODO(), key, "", nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to get secret: %w", err)
+	}
+
+	return *getSecretResponse.Value, nil
+}
+
+// StoreDEK saves new passphrase to Azure key vault.
+func (kms *azureKeyVaultKMS) StoreDEK(key, value string) error {
+	svc, err := kms.getService()
+	if err != nil {
+		return fmt.Errorf("failed to get KMS service: %w", err)
+	}
+
+	setSecretParams := azsecrets.SetSecretParameters{
+		Value: &value,
+	}
+	_, err = svc.SetSecret(context.TODO(), key, setSecretParams, nil)
+	if err != nil {
+		return fmt.Errorf("failed to set seceret %w", err)
+	}
+
+	return nil
+}
+
+// RemoveDEK deletes passphrase from Azure key vault.
+func (kms *azureKeyVaultKMS) RemoveDEK(key string) error {
+	svc, err := kms.getService()
+	if err != nil {
+		return fmt.Errorf("failed to get KMS service: %w", err)
+	}
+
+	_, err = svc.DeleteSecret(context.TODO(), key, nil)
+	if err != nil {
+		responseError, ok := err.(*azcore.ResponseError)
+		if ok && responseError.StatusCode == 404 {
+			return PurgeSecret(svc, key)
```

Comment:
is purging really needed? This example suggests that the key is purged automaticallY: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azkeys#Client.DeleteKey

Reply:
we are using `azsecrets`
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azsecrets#Client.DeleteSecret

Reply:
Soft-delete is fine, right? No need to purge, as it will be purged automatically by the Azure KMS Vault service. The automatic purging seems to be the same for Keys as for Secrets.

Are concerned about creating a new key with the same name before the old key was purged? Is that something that will happen? Secrets are usually identified by the volume-id, which is unique.